core/drivers/riscv_zkr_rng.c

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024 Andes Technology Corporation
+ */
+
+#include <crypto/crypto.h>
+#include <kernel/panic.h>
+#include <riscv.h>
+#include <rng_support.h>
+#include <tee/tee_cryp_utl.h>
+
+#define OPST_BIST 0b00
+#define OPST_WAIT 0b01
+#define OPST_ES16 0b10
+#define OPST_DEAD 0b11
+
+static unsigned long read_seed(void)
+{
+	unsigned long ret = 0;
+
+	/*
+	 * The seed register must be accessed using CSR read-write
+	 * instructions. The write operation is ignored and serves
+	 * to indicate polling and flushing.
+	 */
+	asm volatile("csrrw %0, seed, x0" : "=r" (ret) : : "memory");
+
+	return ret;
+}
+
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
+{
+	uint8_t *ptr = buf;
+	uint32_t val = 0;
+
+	while (len > 0) {
+		val = read_seed();
+
+		switch (val >> 30) {
+		case OPST_BIST:
+		case OPST_WAIT:
+			continue;
+		case OPST_ES16:
+			*ptr++ = val & 0xff;
+			len--;
+			if (len > 0) {
+				*ptr++ = val >> 8;
+				len--;
+			}
+			break;
+		case OPST_DEAD:
+			return TEE_ERROR_BAD_STATE;
+		default:
+			break; /* can't happen */
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+void plat_rng_init(void)
+{
+	bool seed_readable = true;
+	unsigned long tmp = 0;
+
+	/*
+	 * If seed CSR read permission in S-mode is not
+	 * granted by M-mode firmware, OP-TEE will raise
+	 * illegal instruction and jumps to a test trap
+	 * handler below to unset seed_readable.
+	 */
+	asm volatile ("  .align 4\n"
```

Comment:
This is too much as inline assembly, please move it into a `.S` file as a helper function.

Reply:
Moved to `readable_csr()` in core/arch/riscv/include/riscv.h